File: app/javascript/channels/answers_channel.js
import consumer from "./consumer"

consumer.subscriptions.create('AnswersChannel', {
  connected () {
    var cur_path = window.location.pathname;
    if (cur_path.match(/questions\/\d/)) {
      var questionId = [...cur_path.split('/')].pop()
      this.perform('follow', { id: questionId });
    }
  },
  received (data) {
    if(data.event == 'answer created') {
      this.perform('get_answer_html', {answer_id: data.answer_id, question_id: data.question_id, user_id: getCookie('user_id')})
    }
    if(data.event == 'created answer html') {
      $('.answers_list').append(data.answer_html)
    }
  }
})

function getCookie(name) {
  var value = "; " + document.cookie;
  var parts = value.split("; " + name + "=");
  if (parts.length == 2) return parts.pop().split(";").shift();
}

File: app/channels/answers_channel.rb
# frozen_string_literal: true

class AnswersChannel < ApplicationCable::Channel
  def follow(data)
    stream_from "question-#{data['id']}"
  end

  def get_answer_html(data)
    user = if data['user_id'] == 'guest'
             nil
           else
             User.find(data['user_id'])
           end
    answer = Answer.find(data['answer_id'])

    # answer_html = ApplicationController.render(
    #   partial: 'answers/answer',
    #   locals: { answer: answer, user: user }
    # )

    answer_html = ApplicationController.render_with_signed_in_user(
      user,
      partial: 'answers/answer',
      locals: { answer: answer }
    )

    data_back = { event: 'created answer html', answer_html: answer_html }
    transmit data_back
  end
end
